File: ios/ExpoTiktokEventModule.swift
import ExpoModulesCore
import UIKit
import TikTokBusinessSDK
import AppTrackingTransparency
import AdSupport

public class ExpoTiktokEventModule: Module {
  private var isInitialized: Bool = false

  public func definition() -> ModuleDefinition {
    Name("ExpoTiktokEvent")

    Constants([
      "TIKTOK_INIT_STATUS": [
        "SUCCESS": "success",
        "FAIL": "fail"
      ]
    ])

    AsyncFunction("initTikTokSdk") { (appId: String, ttAppId: String) async -> [String: Any] in
      if #available(iOS 14, *) {
          let status = await ATTrackingManager.requestTrackingAuthorization()
          if status != .authorized {
            return [
              "status": "fail",
              "error": [
                "code": "ATT_NOT_AUTHORIZED",
                "message": "User did not authorize ATT"
              ]
            ]
          }
      }

      guard let config = TikTokConfig(appId: appId, tiktokAppId: ttAppId) else {
        return [
          "status": "fail",
          "error": [
            "code": "CONFIG_INIT_ERROR",
            "message": "Failed to create TikTokConfig"
          ]
        ]
      }
      
      return await withCheckedContinuation { continuation in
        TikTokBusiness.initializeSdk(config) { success, error in
          if success {
            self.isInitialized = true
            continuation.resume(returning: [
              "status": "success",
              "error": NSNull()
            ])
          } else {
            continuation.resume(returning: [
              "status": "fail",
              "error": [
                "code": "\(error?._code ?? -1)",
                "message": error?.localizedDescription ?? ""
              ]
            ])
          }
        }
      }
    }
    
    Function("identifyUser") { (externalId: String?, externalUserName: String?, phoneNumber: String?, email: String?) in
      guard self.isInitialized else { return }
      TikTokBusiness.identify(withExternalID: externalId, externalUserName: externalUserName, phoneNumber: phoneNumber, email: email)
    }

    Function("logout") {
      guard self.isInitialized else { return }
      TikTokBusiness.logout()
    }

    Function("trackLoginEvent") {
      guard self.isInitialized else { return }
      let event = TikTokBaseEvent(name:TTEventName.login.rawValue)
      TikTokBusiness.trackTTEvent(event)
    }

    Function("trackRegisterEvent") {
      guard self.isInitialized else { return }
      let event = TikTokBaseEvent(name: TTEventName.registration.rawValue)
      TikTokBusiness.trackTTEvent(event)
    }

    Function("startTrack") {
        guard self.isInitialized else { return }
        TikTokBusiness.setTrackingEnabled(true)
    }
  }
}

File: android/src/main/java/expo/modules/tiktokevent/ExpoTiktokEventModule.kt
package expo.modules.tiktokevent

import android.os.Handler
import android.os.Looper
import android.util.Log
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import com.tiktok.TikTokBusinessSdk;
import com.tiktok.TikTokBusinessSdk.TTConfig;
import com.tiktok.appevents.base.EventName
import com.tiktok.appevents.base.TTBaseEvent

class ExpoTiktokEventModule : Module() {

  private enum class TikTokInitStatus(val value: String) {
    SUCCESS("success"),
    FAIL("fail")
  }

  override fun definition() = ModuleDefinition {

    Name("ExpoTiktokEvent")

    Constants(
      "TIKTOK_INIT_STATUS" to mapOf(
        "SUCCESS" to TikTokInitStatus.SUCCESS.value,
        "FAIL" to TikTokInitStatus.FAIL.value
      )
    )

    AsyncFunction("initTikTokSdk") { appId: String, ttAppId: String, promise: Promise ->
      Handler(Looper.getMainLooper()).post {
        val context = appContext.reactContext?.applicationContext

        if (context == null) {
          promise.resolve(mapOf(
            "status" to TikTokInitStatus.FAIL.value,
            "error" to mapOf(
              "code" to "NO_CONTEXT",
              "message" to "Application context is null"
            )
          ))
          return@post
        }

        val ttConfig = TTConfig(context)
          .setAppId(appId)
          .setTTAppId(ttAppId)
          .setLogLevel(TikTokBusinessSdk.LogLevel.INFO)

        TikTokBusinessSdk.initializeSdk(ttConfig, object : TikTokBusinessSdk.TTInitCallback {
          override fun success() {
            promise.resolve(mapOf(
              "status" to TikTokInitStatus.SUCCESS.value,
              "error" to null
            ))
          }

          override fun fail(code: Int, msg: String?) {
            promise.resolve(mapOf(
              "status" to TikTokInitStatus.FAIL.value,
              "error" to mapOf(
                "code" to code.toString(),
                "message" to msg
              )
            ))
          }
        })
      }
    }

    Function("startTrack") {
      Handler(Looper.getMainLooper()).post {
        TikTokBusinessSdk.startTrack()
      }
    }

    Function("identifyUser") { externalId: String?, externalUserName: String?, phoneNumber: String?, email: String? ->
      Handler(Looper.getMainLooper()).post {
        try {
          TikTokBusinessSdk.identify(
            externalId,
            externalUserName,
            phoneNumber,
            email
          )
        } catch (e: Exception) {
          Log.e("ExpoTiktokEvent", "identify crash: ${e.message}")
        }
      }
    }

    Function("logout") {
      Handler(Looper.getMainLooper()).post {
        TikTokBusinessSdk.logout()
      }
    }

    Function("trackLoginEvent") {
      Handler(Looper.getMainLooper()).post {
        val event = TTBaseEvent.newBuilder(EventName.LOGIN.toString()).build()
        TikTokBusinessSdk.trackTTEvent(event)
      }
    }

    Function("trackRegisterEvent") {
      Handler(Looper.getMainLooper()).post {
        val event = TTBaseEvent.newBuilder(EventName.REGISTRATION.toString()).build()
        TikTokBusinessSdk.trackTTEvent(event)
      }
    }
  }
}